ipykernel/kernelbase.py

```diff
@@ -356,28 +361,91 @@ async def advance_eventloop():
+            while True:
+                msg = await self.shell_socket.recv_multipart()
+
+                # Deserialize whole message just to get subshell_id.
+                # Keep original message to send to subshell_id unmodified.
+                # Ideally only want to deserialize message once.
+                copy = not isinstance(msg[0], zmq.Message)
+                _, msg2 = self.session.feed_identities(msg, copy=copy)
+                try:
+                    msg3 = self.session.deserialize(msg2, content=False, copy=copy)
+                    subshell_id = msg3["header"].get("subshell_id")
+
+                    # Find inproc pair socket to use to send message to correct subshell.
+                    socket = self.shell_channel_thread.manager.get_shell_channel_socket(subshell_id)
+                    assert socket is not None
+                    socket.send_multipart(msg, copy=False)
+                except Exception:
+                    self.log.error("Invalid message", exc_info=True)  # noqa: G201
+        except BaseException as e:
+            if self.shell_stop.is_set():
+                return
+            raise e
+
+    async def shell_main(self, subshell_id: str | None):
+        """Main loop for a single subshell."""
+        if self._supports_kernel_subshells():
+            if subshell_id is None:
+                assert threading.current_thread() == threading.main_thread()
+            else:
+                assert threading.current_thread() not in (
+                    self.shell_channel_thread,
+                    threading.main_thread(),
+                )
+            # Inproc pair socket that this subshell uses to talk to shell channel thread.
+            socket = self.shell_channel_thread.manager.get_other_socket(subshell_id)
+        else:
+            assert subshell_id is None
+            assert threading.current_thread() == threading.main_thread()
+            socket = self.shell_socket
+
         async with create_task_group() as tg:
-            tg.start_soon(self.process_shell)
-            await to_thread.run_sync(self.shell_stop.wait)
-            tg.cancel_scope.cancel()
+            tg.start_soon(self.process_shell, socket)
+            if subshell_id is None:
+                # Main subshell.
+                await to_thread.run_sync(self.shell_stop.wait)
+                tg.cancel_scope.cancel()
 
-    async def process_shell(self):
+    async def process_shell(self, socket=None):
         try:
             while True:
-                await self.process_shell_message()
+                await self.process_shell_message(socket=socket)
         except BaseException as e:
             if self.shell_stop.is_set():
                 return
             raise e
 
-    async def process_shell_message(self, msg=None):
-        assert self.shell_socket is not None
+    async def process_shell_message(self, msg=None, socket=None):
+        if not socket:
+            socket = self.shell_socket
```

Comment:
I'm going to assume later will make socket mandatory ?
If so we may want to put a comment or a FutureWarning ?

Reply:
An input argument of `socket=None` is correct if subshells are not supported in which case the behaviour of this function is unchanged compared to the `main` branch, which is (probably) important for downstream libraries. There are a few situations where this is the case, I will add extra comments to explain. Also, here I will move these lines to the `_supports_kernel_subshells` block below to assert that `socket` can only be `None` if subshells are not supported.

Reply:
I believe it is ok if we want to think about making `socket=` mandatory in a few years. 

I tend to not like API where things are sometime magic and sometime not.